### DISCUSSION TITLE
[Question] Chaînes de caractères.

### DISCUSSION THREAD

**Discussion by pierrecamilleri** [publicodes/publicodes#414] [General]
Bonjour à tous, 

J'ai plusieurs questions au sujet du type chaînes de caractères.

- D'abord, j'ai l'impression que la syntaxe `'"xxx"'` n'est pas documentée (mais utilisée [ici](https://publi.codes/docs/principes-de-base#d%C3%A9sactivation-de-branche) dans la documentation pour un sujet sans lien). Je l'ai découvert en lisant du code. Est-ce que vous accepteriez une PR pour le rajouter à la documentation (à condition que je comprenne bien comment ça fonctionne ^^') ?  
- Ensuite, je trouve le comportement contre-intuitif, je m'interroge pourquoi à la définition d'une variable, il faut doubler les quotes (`var: '"abc"'`) pour que ce ne soit pas interprété comme une référence à une autre règle, alors que dans le cas d'une comparaison, il ne faut pas (`var = "abc"`).  J'ai l'impression que dans le cas de la définition de variable, les quotes sont tout simplement ignorées (par exemple : `code NAF: "12.34Z"` va être interprété comme 12,34 Z - oui, je me suis fait mordre gentiment, ça ne renvoyait pas d'erreur mais ça ne faisait pas non plus ce que je voulais).

EDIT: je rajouterais que ce n'est pas non plus complètement intuitif sur comment utiliser la syntaxe publicodes `'"xxx"'` en injectant des données avec `Engine.setSituation` de l'API Core. Si quelqu'un s'interroge, la bonne manière est `'"xxx"'` et non `` `'"xxx"'` `` (ambiguité de ce qu'est une "expression publicodes bien formée.")

**Comment by johangirod** [comment]
Salut Pierre,

 Effectivement, la syntaxe double quote `'"xxx'"` n'est pas documentée explicitement. Tu es bien évidément plus que le bienvenu pour faire une PR qui précise cela dans la doc ! 

Cela est du au fait que les projets publicodes utilisent un parser `yaml`. En `yaml` les syntaxes suivantes ont le même sens et aboutissent au même objet parsé :
```yaml
a: b
a: "b"
```
Il faut donc échapper le caractèrez `"` pour qu'il soit « transmis » au parser publicodes. Le plus simple est d'utiliser le double quote (mais on peut aussi imaginer le faire avec `|` ou d'autres syntaxes yaml. 

En revanche, lorsque l'on écrit :
```yaml
a: var = "abc"
```
Le parser yaml échappe automatiquement les quote `"`, il n'y a pas besoin de le faire manuellement.

Plus en détail, en l'état actuel, il y a plusieurs représentations de code `publicodes` qui co-existent. 

1. une en yaml, qui est celle utilisée par la plupart des projets pour écrire les règles
2. une en objets JavaScript qui est la « source », celle qui est réellement parsée par l'Engine
3. la dernière, l'AST, qui est celle utilisée pour l'évaluation

Nous souhaitons faire en sorte de fusionner 1 et 2 pour le futur en écrivant notre propre parser. La syntaxe officielle serait alors celle inspirée du yaml.

**Reply by pierrecamilleri** [reply]
Ok super, merci pour ces explications détaillées, je comprends mieux, et ça éclaire également la forme d'objet attendue en javascript. 

Je vous propose un paragraphe de documentation sous peu alors !
